pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexQualifiedName.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.ast;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+
+import net.sourceforge.pmd.lang.ast.QualifiedName;
+
+import apex.jorje.semantic.symbol.type.TypeInfo;
+
+/**
+ * Qualified name of an apex class or method.
+ *
+ * @author Clément Fournier
+ */
+public class ApexQualifiedName implements QualifiedName {
+
+    private static final Pattern FORMAT = Pattern.compile("(\\w+)__(\\w+)(.(\\w+))?(#(\\w+))?"); // TODO
+
+    private final String nameSpace;
+    private final String[] classes;
+    private final String operation;
+
+
+    private ApexQualifiedName(String nameSpace, String[] classes, String operation) {
+        this.nameSpace = nameSpace;
+        this.operation = operation;
+        this.classes = classes;
+    }
+
+
+
+    public String getOperation() {
+        return operation;
+    }
+
+
+
+    public String[] getClasses() {
+        return Arrays.copyOf(classes, classes.length);
+    }
+
+
+    /**
+     * Gets the namespace prefix of this resource.
+     *
+     * @return The namespace prefix
+     */
+    public String getNameSpace() {
+        return nameSpace;
+    }
+
+
+    @Override
+    public boolean isClass() {
+        return operation == null;
+    }
+
+
+    @Override
+    public boolean isOperation() {
+        return operation != null;
+    }
+
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(nameSpace).append("__");
+        sb.append(classes[0]);
+
+        if (classes.length > 1) {
+            sb.append('.').append(classes[1]);
+        }
+
+        if (isOperation()) {
+            sb.append("#").append(operation);
+        }
+
+        return sb.toString();
+    }
+
+
+    @Override
+    public ApexQualifiedName getClassName() {
+        if (isClass()) {
+            return this;
+        }
+
+        return new ApexQualifiedName(this.nameSpace, this.classes, null);
+    }
+
+
+    @Override
+    public int hashCode() {
+        int result = nameSpace.hashCode();
+        result = 31 * result + Arrays.hashCode(classes);
+        result = 31 * result + (operation != null ? operation.hashCode() : 0);
+        return result;
+    }
+
+
+    @Override
+    public boolean equals(Object obj) {
+        return obj instanceof ApexQualifiedName
+            && Objects.deepEquals(classes, ((ApexQualifiedName) obj).classes)
+            && Objects.equals(operation, ((ApexQualifiedName) obj).operation)
+            && Objects.equals(nameSpace, ((ApexQualifiedName) obj).nameSpace);
+
+    }
+
+
+    public static ApexQualifiedName ofString(String toParse) {
+        throw new UnsupportedOperationException();
```

Comment:
I don't really understand it as well :s I figured we could add the namespace in the qualified name, even though I guess for most projects, all code belong in the same namespace, as it's apparently defined by organisation
Namespace might only be relevant when resolving method usages, i don't know at this point